### PR TITLE
Handle noop coalescing for counting if client should join write

### LINKED ISSUE #5628: Logic to calculate if client has pending non-acked ops is not correct
This logic in DeltaManager is incorrect, as it does not take into account noop coalescing:
```
        // if we have any non-acked ops from last connection, reconnect as "write".
        // without that we would connect in view-only mode, which will result in immediate
        // firing of "connected" event from Container and switch of current clientId (as tracked
        // by all DDSes). This will make it impossible to figure out if ops actually made it through,
        // so DDSes will immediately resubmit all pending ops, and some of them will be duplicates, corrupting document
        if (this.clientSequenceNumberObserved !== this.clientSequenceNumber) {
            requestedMode = "write";
        }
```

Please note that it will moved with @jatgarg's PR https://github.com/microsoft/FluidFramework/pull/5484/files around, but it will stay the same (in terms of logic) and needs fixing.

### PR DESCRIPTION
**Pull request by jatgarg**
Fixes: https://github.com/microsoft/FluidFramework/issues/5628

### CODE REVIEW

File: packages/loader/container-loader/src/deltaManager.ts [hunk 1]
```diff
@@ -203,6 +203,8 @@ export class DeltaManager
     private connection: IDocumentDeltaConnection | undefined;
     private clientSequenceNumber = 0;
     private clientSequenceNumberObserved = 0;
+    // Represents the latest topmost noops sent by the client.
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:206]: I'd change naming / description (it was not obvious for me what it is until I looked at the rest of implementation).

File: packages/loader/container-loader/src/deltaManager.ts [hunk 2]
```diff
@@ -357,7 +359,8 @@ export class DeltaManager
     }
 
     public shouldJoinWrite(): boolean {
-        return this.clientSequenceNumber !== this.clientSequenceNumberObserved;
+        // We don't have to wait for ack for topmost NoOps. So subtract those.
+        return (this.clientSequenceNumber - this.lastNoOps) !== this.clientSequenceNumberObserved;
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:363]: !== will not work here. Consider two cases:
- noops were sent but not received at all by this client (for any reason)
- noops were sent, but some of them were acked (sequenced and received back).

These two cases will produce same values for thisclientSequenceNumber & lastNoOps, but different for this.clientSequenceNumberObserved.
Thus it should be <=

File: packages/loader/container-loader/src/deltaManager.ts [hunk 3]
```diff
@@ -772,6 +775,12 @@ export class DeltaManager
             type,
         };
 
+        if (type === MessageType.NoOp) {
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:778]: I think (though I'd double check) we can be more generic here, and check for all system ops (isSystemMessage).
90% of them should not be sent by client, but noop & summarize are in the list.
I think it's good change, given that Container.submitContainerMessage() validates that container can only sent 3 types of messages (i.e. should not be able to sent system ops).

Comment (jatgarg) [reply packages/loader/container-loader/src/deltaManager.ts:778]: I think it is more about what ops server may not acknowledge and that are noops. Server should ack the summary op and if it is not acked then client should resend it. So I think noop check is fine?

File: packages/loader/container-loader/src/deltaManager.ts (hunk 2 above)
Comment (jatgarg) [reply packages/loader/container-loader/src/deltaManager.ts]: Yes, should be "<".

File: packages/loader/container-loader/src/deltaManager.ts (hunk 3 above)
Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:778]: Can you please check if we indeed resend summary Op on reconnection if it was lost?
If we do, then yes, I agree with you (even though the point is a but murky, as another container runtime may have different behavior)

File: packages/loader/container-loader/src/deltaManager.ts [hunk 4]
```diff
@@ -203,6 +203,8 @@ export class DeltaManager
     private connection: IDocumentDeltaConnection | undefined;
     private clientSequenceNumber = 0;
     private clientSequenceNumberObserved = 0;
+    // Counts the number of noops sent by the client which may not be acked.
+    private topNoOpsCount = 0;
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:207]: topNoOpsCount -> noopCount ? Or maybe trailingNoopCount?

**Review by vladsud — APPROVED** [review]

File: packages/loader/container-loader/src/deltaManager.ts (hunk 4 above)
Comment (jatgarg) [reply packages/loader/container-loader/src/deltaManager.ts]: trailingNoopCount seems good.

File: packages/loader/container-loader/src/deltaManager.ts (hunk 3 above)
Comment (jatgarg) [reply packages/loader/container-loader/src/deltaManager.ts:778]: As I checked summarizer never rejoins on disconnection. So this won't occur for summarize op.

Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts:778]: This might change in future. Either way - it's fine one way or another :)

**Pull request merged**